File: native/aloud-tts/ios/AloudTtsModule.swift
import AVFoundation
import Foundation
// React ships as a prebuilt Swift-consumable framework module in modern RN
// (0.79+ / React-Core-prebuilt): `RCTEventEmitter` and friends come from
// `import React`, not a bridging header. The C-only Rust ABI is exposed the
// same way, as a Clang module — see AloudCore.swift's `import AloudCoreFFI`
// and core/include/module.modulemap.
import React
import UIKit

/// The iOS native module. It owns three things and keeps them in step:
///   1. the shared Rust core (`AloudCore`) — the source of truth for position,
///   2. `AVSpeechSynthesizer` — the platform TTS engine,
///   3. `AVAudioSession` — the shared audio hardware, which we must share
///      politely with VoiceOver.
///
/// JS sends coarse intents; this module runs the tight word-boundary loop
/// natively and streams `Snapshot`s back up. See `AloudTtsSpec.ts` for why the
/// loop lives here and not in JS.
@objc(AloudTts)
final class AloudTtsModule: RCTEventEmitter {
    private let synthesizer = AVSpeechSynthesizer()
    private var core: AloudCore?
    private var hasListeners = false
    private var activeTraceId = "-"
    /// AVSpeechUtterance.rate to apply to future utterances (see `setRate`).
    private var currentRate = AVSpeechUtteranceDefaultSpeechRate
    /// The exact `AVSpeechUtterance` instance we last told the synthesizer to
    /// speak. Every delegate callback below checks the callback's `utterance`
    /// against this by REFERENCE before touching the core.
    ///
    /// Why this matters: seeking (tap-to-seek, next/prev) stops the in-flight
    /// utterance and immediately starts a new one. `stopSpeaking` does not
    /// guarantee its delegate callback for the OLD utterance is suppressed
    /// before the NEW one starts producing callbacks of its own — on a fast
    /// double-seek this could otherwise feed a stale word-boundary report from
    /// the utterance we just abandoned into the core, which would move the
    /// highlight to the wrong word for a moment. Without this guard that shows
    /// up as an intermittent "highlight desyncs from the audio" bug — it only
    /// reproduces when the race is lost, which is exactly why it was
    /// inconsistent to reproduce by hand.
    private var currentUtterance: AVSpeechUtterance?

    override init() {
        super.init()
        synthesizer.delegate = self
        // React to VoiceOver turning on/off so we can coordinate the audio
        // session (see `configureAudioSession`).
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(voiceOverStatusChanged),
            name: UIAccessibility.voiceOverStatusDidChangeNotification,
            object: nil
        )
    }

    // RCTEventEmitter plumbing -------------------------------------------------

    override static func requiresMainQueueSetup() -> Bool { true }
    override func supportedEvents() -> [String]! { ["AloudSnapshot"] }
    override func startObserving() { hasListeners = true }
    override func stopObserving() { hasListeners = false }

    // Exported intents ---------------------------------------------------------

    @objc(getCoreVersion:rejecter:)
    func getCoreVersion(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(AloudCore.version)
    }

    @objc(load:traceId:resolver:rejecter:)
    func load(_ text: String, traceId: String,
              resolver resolve: RCTPromiseResolveBlock,
              rejecter reject: RCTPromiseRejectBlock) {
        activeTraceId = traceId
        do {
            let core = try AloudCore(text: text)
            self.core = core
            log("load", "unitCount=\(core.unitCount)")
            let snap = try core.dispatch(.getState)
            resolve(snap.asBridgePayload())
        } catch {
            reject("load_failed", "\(error)", error)
        }
    }

    @objc(play:resolver:rejecter:)
    func play(_ traceId: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        run(traceId, resolve, reject) { core in
            let snap = try core.dispatch(.play)
            self.configureAudioSession(active: true)
            self.speakCurrentUtterance(snap)
            return snap
        }
    }

    @objc(pause:resolver:rejecter:)
    func pause(_ traceId: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        run(traceId, resolve, reject) { core in
            self.synthesizer.stopSpeaking(at: .word)
            self.currentUtterance = nil
            let snap = try core.dispatch(.pause)
            self.configureAudioSession(active: false)
            return snap
        }
    }

    @objc(next:resolver:rejecter:)
    func next(_ traceId: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        run(traceId, resolve, reject) { core in
            self.synthesizer.stopSpeaking(at: .immediate)
            let snap = try core.dispatch(.next)
            self.speakCurrentUtterance(snap)
            return snap
        }
    }

    @objc(prev:resolver:rejecter:)
    func prev(_ traceId: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        run(traceId, resolve, reject) { core in
            self.synthesizer.stopSpeaking(at: .immediate)
            let snap = try core.dispatch(.prev)
            self.speakCurrentUtterance(snap)
            return snap
        }
    }

    @objc(seekUnit:traceId:resolver:rejecter:)
    func seekUnit(_ unit: NSNumber, traceId: String,
                  resolver resolve: RCTPromiseResolveBlock,
                  rejecter reject: RCTPromiseRejectBlock) {
        run(traceId, resolve, reject) { core in
            self.synthesizer.stopSpeaking(at: .immediate)
            let snap = try core.dispatch(.seekUnit(unit: unit.intValue))
            self.speakCurrentUtterance(snap)
            return snap
        }
    }

    @objc(seekByte:traceId:resolver:rejecter:)
    func seekByte(_ byte: NSNumber, traceId: String,
                  resolver resolve: RCTPromiseResolveBlock,
                  rejecter reject: RCTPromiseRejectBlock) {
        run(traceId, resolve, reject) { core in
            self.synthesizer.stopSpeaking(at: .immediate)
            let snap = try core.dispatch(.seekByte(byte: byte.intValue))
            self.speakCurrentUtterance(snap)
            return snap
        }
    }

    /// Set the speech-rate multiplier (1.0 = normal). We map it onto Apple's
    /// rate scale around its default and clamp to the platform's min/max. The new
    /// rate applies to the next utterance; we do not restart the current one.
    @objc(setRate:traceId:resolver:rejecter:)
    func setRate(_ rate: NSNumber, traceId: String,
                 resolver resolve: RCTPromiseResolveBlock,
                 rejecter reject: RCTPromiseRejectBlock) {
        activeTraceId = traceId
        let mapped = AVSpeechUtteranceDefaultSpeechRate * rate.floatValue
        currentRate = min(AVSpeechUtteranceMaximumSpeechRate,
                          max(AVSpeechUtteranceMinimumSpeechRate, mapped))
        log("setRate", "multiplier=\(rate.floatValue) rate=\(currentRate)")
        resolve(nil)
    }

    @objc(release:rejecter:)
    func release(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        synthesizer.stopSpeaking(at: .immediate)
        currentUtterance = nil
        configureAudioSession(active: false)
        core = nil // AloudCore.deinit frees the Rust session
        resolve(nil)
    }

    // Internals ----------------------------------------------------------------

    private func run(_ traceId: String,
                     _ resolve: RCTPromiseResolveBlock,
                     _ reject: RCTPromiseRejectBlock,
                     _ body: (AloudCore) throws -> Snapshot) {
        activeTraceId = traceId
        guard let core = core else {
            reject("no_session", "load() must be called first", nil)
            return
        }
        do {
            let snap = try body(core)
            resolve(snap.asBridgePayload())
        } catch {
            log("error", "\(error)")
            reject("dispatch_failed", "\(error)", error)
        }
    }

    /// Speak the current sentence (or the remaining suffix of it after a
    /// mid-sentence seek — see `Snapshot.utterance`) if we are in the playing
    /// state.
    private func speakCurrentUtterance(_ snap: Snapshot) {
        guard snap.status == "playing", !snap.utterance.isEmpty else {
            currentUtterance = nil
            return
        }
        let utterance = AVSpeechUtterance(string: snap.utterance)
        utterance.voice = AVSpeechSynthesisVoice(language: "en-US")
        utterance.rate = currentRate
        currentUtterance = utterance
        synthesizer.speak(utterance)
    }

    /// Configure the shared audio session. `.spokenAudio` mode is the key
    /// screen-reader-friendly choice: it tells the system this is speech, so it
    /// mixes sensibly with VoiceOver rather than ducking it into inaudibility.
    private func configureAudioSession(active: Bool) {
        let session = AVAudioSession.sharedInstance()
        do {
            if active {
                try session.setCategory(.playback, mode: .spokenAudio, options: [.duckOthers])
                try session.setActive(true)
            } else {
                // Yield the session so VoiceOver/other apps regain full volume.
                try session.setActive(false, options: [.notifyOthersOnDeactivation])
            }
        } catch {
            log("audio_session", "failed: \(error)")
        }
    }

    @objc private func voiceOverStatusChanged() {
        // If VoiceOver turns on mid-playback, pause so the two speech streams do
        // not talk over each other; the user can resume deliberately.
        if UIAccessibility.isVoiceOverRunning, synthesizer.isSpeaking {
            synthesizer.pauseSpeaking(at: .word)
            if let core = core, let snap = try? core.dispatch(.pause) {
                emit(snap)
            }
        }
    }

    private func emit(_ snap: Snapshot) {
        guard hasListeners else { return }
        sendEvent(withName: "AloudSnapshot", body: snap.asBridgePayload())
    }

    private func log(_ op: String, _ message: String) {
        NSLog("[aloud t=\(activeTraceId) layer=native op=\(op)] \(message)")
    }
}

// MARK: - AVSpeechSynthesizerDelegate — the native word-boundary loop

extension AloudTtsModule: AVSpeechSynthesizerDelegate {
    /// Word boundary. `characterRange.location` is a **UTF-16** offset into the
    /// utterance string — exactly what the core's `WordBoundary` command wants.
    /// We feed it straight in and stream the resulting highlight to JS. No
    /// offset math happens here; that lives once, in Rust.
    ///
    /// The `utterance === currentUtterance` guard discards callbacks for an
    /// utterance we've since abandoned (seeked/stopped away from) — see the
    /// doc comment on `currentUtterance` for why this race is real.
    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                           willSpeakRangeOfSpeechString characterRange: NSRange,
                           utterance: AVSpeechUtterance) {
        guard utterance === currentUtterance, let core = core else { return }
        if let snap = try? core.dispatch(.wordBoundary(utf16Offset: characterRange.location)) {
            emit(snap)
        }
    }

    /// Sentence finished. Advance the core; if it produced another sentence and
    /// we are still playing, speak it — this is the auto-advance loop.
    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                           didFinish utterance: AVSpeechUtterance) {
        guard utterance === currentUtterance, let core = core else { return }
        guard let snap = try? core.dispatch(.next) else { return }
        emit(snap)
        if snap.status == "playing" {
            speakCurrentUtterance(snap)
        } else {
            // Document finished: release the audio session politely.
            configureAudioSession(active: false)
        }
    }
}

File: native/aloud-tts/android/src/main/java/com/aloud/tts/AloudTtsModule.kt
package com.aloud.tts

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.os.Build
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.view.accessibility.AccessibilityManager
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.util.Locale

/**
 * The Android native module. Mirror image of the iOS one: it owns the shared
 * Rust core, the platform `TextToSpeech` engine, and audio focus, and runs the
 * word-boundary loop natively while streaming `Snapshot`s to JS.
 *
 * The per-vendor quirk this file exists to absorb: Android reports word
 * boundaries via `UtteranceProgressListener.onRangeStart`, whose `start` is a
 * **UTF-16** code-unit index into the utterance — the same shape iOS gives us,
 * and exactly what the core's `WordBoundary` command consumes.
 */
class AloudTtsModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var tts: TextToSpeech? = null
    private var core: AloudCore? = null
    private var ready = false
    private var activeTraceId = "-"
    private var focusRequest: AudioFocusRequest? = null
    /**
     * The utteranceId of the most recent `speak()` call. `onRangeStart`/`onDone`
     * below discard any callback whose id doesn't match this.
     *
     * Why: seeking (tap-to-seek, next/prev) stops the in-flight utterance and
     * immediately starts a new one. `TextToSpeech.stop()` does not guarantee a
     * callback already queued for the OLD utterance is suppressed before the NEW
     * one starts producing its own — without this guard, a stale word-boundary
     * report from the utterance we just abandoned could move the highlight to
     * the wrong word for a moment (an intermittent "highlight desyncs from the
     * audio" bug). The id must be unique per `speak()` call, not just per
     * sentence — a seek that lands in the SAME sentence still starts a genuinely
     * new utterance, so `"aloud-$unit"` alone can't tell old and new apart.
     */
    private var currentUtteranceId: String? = null
    private var utteranceGeneration = 0

    private val audioManager =
        reactContext.getSystemService(Context.AUDIO_SERVICE) as AudioManager

    override fun getName() = "AloudTts"

    // --- exported intents ----------------------------------------------------

    @ReactMethod
    fun getCoreVersion(promise: Promise) = promise.resolve(AloudCore.version)

    @ReactMethod
    fun load(text: String, traceId: String, promise: Promise) {
        activeTraceId = traceId
        try {
            core?.release()
            core = AloudCore(text)
            ensureEngine {
                val snap = core!!.dispatch(AloudCore.getState())
                promise.resolve(snap.toWritableMap())
            }
        } catch (e: CoreException) {
            promise.reject(e.code, e.message)
        }
    }

    @ReactMethod
    fun play(traceId: String, promise: Promise) = run(traceId, promise) { core ->
        val snap = core.dispatch(AloudCore.play())
        requestAudioFocus()
        speakCurrent(snap)
        snap
    }

    @ReactMethod
    fun pause(traceId: String, promise: Promise) = run(traceId, promise) { core ->
        tts?.stop()
        currentUtteranceId = null
        val snap = core.dispatch(AloudCore.pause())
        abandonAudioFocus()
        snap
    }

    @ReactMethod
    fun next(traceId: String, promise: Promise) = run(traceId, promise) { core ->
        tts?.stop()
        val snap = core.dispatch(AloudCore.next())
        speakCurrent(snap)
        snap
    }

    @ReactMethod
    fun prev(traceId: String, promise: Promise) = run(traceId, promise) { core ->
        tts?.stop()
        val snap = core.dispatch(AloudCore.prev())
        speakCurrent(snap)
        snap
    }

    @ReactMethod
    fun seekUnit(unit: Int, traceId: String, promise: Promise) = run(traceId, promise) { core ->
        tts?.stop()
        val snap = core.dispatch(AloudCore.seekUnit(unit))
        speakCurrent(snap)
        snap
    }

    /**
     * Set the speech-rate multiplier (1.0 = normal). Android's `setSpeechRate`
     * already uses the same convention, so we pass it through (clamped) and it
     * applies to subsequent utterances.
     */
    @ReactMethod
    fun setRate(rate: Double, traceId: String, promise: Promise) {
        activeTraceId = traceId
        val clamped = rate.coerceIn(0.5, 2.0).toFloat()
        tts?.setSpeechRate(clamped)
        log("setRate", "rate=$clamped")
        promise.resolve(null)
    }

    @ReactMethod
    fun seekByte(byte: Int, traceId: String, promise: Promise) = run(traceId, promise) { core ->
        tts?.stop()
        val snap = core.dispatch(AloudCore.seekByte(byte))
        speakCurrent(snap)
        snap
    }

    @ReactMethod
    fun release(promise: Promise) {
        tts?.stop()
        currentUtteranceId = null
        tts?.shutdown()
        tts = null
        abandonAudioFocus()
        core?.release()
        core = null
        promise.resolve(null)
    }

    // RN requires these for NativeEventEmitter; no-op is fine.
    @ReactMethod fun addListener(eventName: String) {}
    @ReactMethod fun removeListeners(count: Int) {}

    // --- internals -----------------------------------------------------------

    private inline fun run(traceId: String, promise: Promise, body: (AloudCore) -> Snapshot) {
        activeTraceId = traceId
        val core = this.core ?: return promise.reject("no_session", "load() must be called first")
        try {
            promise.resolve(body(core).toWritableMap())
        } catch (e: CoreException) {
            promise.reject(e.code, e.message)
        }
    }

    private fun ensureEngine(onReady: () -> Unit) {
        if (ready) return onReady()
        tts = TextToSpeech(reactContext) { status ->
            if (status == TextToSpeech.SUCCESS) {
                tts?.language = Locale.US
                tts?.setOnUtteranceProgressListener(progressListener)
                ready = true
                onReady()
            } else {
                log("engine", "TextToSpeech init failed: $status")
            }
        }
    }

    private fun speakCurrent(snap: Snapshot) {
        if (snap.status != "playing" || snap.utterance.isEmpty()) {
            currentUtteranceId = null
            return
        }
        // A unique-per-call id (not just per sentence) lets onDone/onRangeStart
        // tell this utterance apart from one we've since abandoned mid-sentence.
        utteranceGeneration++
        val id = "aloud-${snap.unit}-$utteranceGeneration"
        currentUtteranceId = id
        tts?.speak(snap.utterance, TextToSpeech.QUEUE_FLUSH, null, id)
    }

    private val progressListener = object : UtteranceProgressListener() {
        /** UTF-16 code-unit index into the utterance — fed straight to the core. */
        override fun onRangeStart(utteranceId: String, start: Int, end: Int, frame: Int) {
            if (utteranceId != currentUtteranceId) return
            val core = core ?: return
            runCatching { core.dispatch(AloudCore.wordBoundary(start)) }
                .onSuccess { emit(it) }
        }

        override fun onDone(utteranceId: String) {
            if (utteranceId != currentUtteranceId) return
            val core = core ?: return
            runCatching { core.dispatch(AloudCore.next()) }.onSuccess { snap ->
                emit(snap)
                if (snap.status == "playing") speakCurrent(snap) else abandonAudioFocus()
            }
        }

        @Deprecated("Deprecated in Java", ReplaceWith(""))
        override fun onError(utteranceId: String) {
            log("tts", "utterance error: $utteranceId")
        }

        override fun onStart(utteranceId: String) {}
    }

    private fun talkBackActive(): Boolean {
        val am = reactContext.getSystemService(Context.ACCESSIBILITY_SERVICE)
                as? AccessibilityManager ?: return false
        return am.isTouchExplorationEnabled
    }

    /** Request media/assistant focus. Marked as SPEECH content so the system
     *  coordinates with TalkBack's accessibility stream sensibly. */
    private fun requestAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val attrs = AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_ASSISTANT)
                .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                .build()
            val request = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN)
                .setAudioAttributes(attrs)
                .setWillPauseWhenDucked(true)
                .build()
            focusRequest = request
            audioManager.requestAudioFocus(request)
        }
    }

    private fun abandonAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            focusRequest?.let { audioManager.abandonAudioFocusRequest(it) }
            focusRequest = null
        }
    }

    private fun emit(snap: Snapshot) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("AloudSnapshot", snap.toWritableMap())
    }

    private fun log(op: String, message: String) =
        android.util.Log.i("Aloud", "[aloud t=$activeTraceId layer=native op=$op] $message")
}

private fun Snapshot.toWritableMap(): WritableMap {
    val map = Arguments.createMap()
    map.putString("status", status)
    map.putInt("unit", unit)
    map.putInt("unitCount", unitCount)
    map.putInt("token", token)
    map.putInt("tokenCount", tokenCount)
    map.putString("utterance", utterance)
    if (highlight != null) {
        val h = Arguments.createMap()
        h.putInt("start", highlight.start)
        h.putInt("end", highlight.end)
        map.putMap("highlight", h)
    } else {
        map.putNull("highlight")
    }
    return map
}
